src/boxes/mp4_box.rs

```diff
@@ -0,0 +1,127 @@
+/*
+ * © 2022 Arastoo Bozorgi
+ * All rights reserved.
+ */
+
+use crate::Error;
+use downcast_rs::Downcast;
+
+pub trait Mp4Box: Downcast {
+    /// Read the atom from the data and parse it
+    fn parse(data: &[u8], start: usize, level: u8) -> Result<Self, Error>
+    where
+        Self: Sized;
+
+    /// The start address of the box
+    fn start(&self) -> usize;
+
+    /// The end address of the box
+    fn end(&self) -> usize;
+
+    /// The box size in bytes
+    fn size(&self) -> usize;
+
+    /// The box name
+    fn name(&self) -> &str;
+
+    /// Read the box content
+    fn read(&self) -> Result<Vec<u8>, Error>;
+
+    /// Get the internal boxes of this box
+    fn fields(&self) -> Option<Vec<&Box<dyn Mp4Box>>>;
+
+    /// The box level
+    fn level(&self) -> u8;
+}
+
+impl_downcast!(Mp4Box);
+
+impl std::fmt::Debug for dyn Mp4Box {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "{}: {{start address: {}, size: {}}} \n",
+            self.name(),
+            self.start(),
+            self.size()
+        )?;
+
+        let internals = self.fields();
+        if internals.is_some() {
```

Comment:
warning: called `unwrap` on `internals` after checking its variant with `is_some`
  --> src\boxes\mp4_box.rs:51:29
   |
50 |         if internals.is_some() {
   |         ---------------------- help: try: `if let Some(..) = internals`
51 |             for internal in internals.unwrap() {
   |                             ^^^^^^^^^^^^^^^^^^
   |
   = note: `#[warn(clippy::unnecessary_unwrap)]` on by default
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unnecessary_unwrap